### ISSUE TITLE
Hidden Markov Model - SP500

### ISSUE THREAD

**Issue by kevinostoll-dfg** [deerfieldgreen/yfinance-scaling-system#1]
# Implement a universal HMM for all algos to reference 

## reference material
- https://www.mdpi.com/1911-8074/13/12/311
- https://www.quantconnect.com/docs/v2/research-environment/applying-research/hidden-markov-models
- https://www.quantstart.com/articles/market-regime-detection-using-hidden-markov-models-in-qstrader/
- https://www.quantstart.com/articles/hidden-markov-models-for-regime-detection-using-r/

**Comment by kevinostoll-dfg** [comment]
Not only do we want a regime detection but a "for how long has it been this way" score or label.

sp500
nasdaq

time window - date, symbol
A B C

days - duration
mean duration of state changes
From A to B
From B to C
From C to B

other ....

**Comment by kevinostoll-dfg** [comment]
- data comes from yfinance, based on the last X days from now. You can modify this value. For testing it's 45-60 day for speed.
- Added code which calculates HMM per day and logs it to a data frame
- Calculates 10day MSE
- Added feature scaling 
- HMM model fit

## Next Steps
1. From the dataframe, derive a "duration of regime" metric
2. Get peer reviewed 
3. Make changes / improvements to the model



### Section 3
I focused a lot on Section 3 of the paper for this draft: 
4. Model Specification
3.1. Data Source and Processing

**Comment by kevinostoll-dfg** [comment]
- added code to track changes in regime. 
it's not working entirely properly, but it's close.


```


# --- Track Regime Changes ---
def track_regime_changes(df):
    regime_changes = []
    current_regime = None
    start_date = None
    for i in range(len(df)):
        regime = df["Regime"].iloc[i]
        date = df["Date"].iloc[i]
        if regime != current_regime:
            if current_regime is not None:
                regime_changes.append(
                    {
                        "Start_Date": start_date,
                        "End_Date": df["Date"].iloc[i - 1],
                        "Regime": current_regime,
                        "Duration_Days": (df["Date"].iloc[i - 1] - start_date).days,
                    }
                )
            current_regime = regime
            start_date = date
    # Add the last regime change
    if current_regime is not None:
        regime_changes.append(
            {
                "Start_Date": start_date,
                "End_Date": df["Date"].iloc[-1],
                "Regime": current_regime,
                "Duration_Days": (df["Date"].iloc[-1] - start_date).days,
            }
        )
    return pd.DataFrame(regime_changes)

regime_changes_df = track_regime_changes(results_df)

```

**Comment by kevinostoll-dfg** [comment]
utils/mkt-regime - repository moved :)

**Comment by kevinostoll-dfg** [comment]
@howardhou-dfg - making progress on a HMM regime detection system


- [ ] Add k-means clustering to the features
K-Means Clustering: Use k-means clustering on your features to get initial estimates of the means for each hidden state. This makes the initialization less sensitive to your specific range choices.

- [ ] How do I test this?
- [ ] Probably need to bring the code into a Colab notebook for visuals




- added KS test to the code 
```
Regime Detection and Classification:
Kolmogorov-Smirnov Test: The use of the KS test is innovative, but it might not be the best approach for regime classification. The HMM already provides probabilities for each state; use these probabilities directly instead of trying to fit other distributions.
Probability Thresholds: The thresholds (0.4 and 0.1) for regime detection are arbitrary. Use cross-validation to determine optimal thresholds.
```

- added Log Volume and Momentum to the feature set 
```

# Example: Adding momentum (e.g., past month return) and a value proxy
data['Momentum'] = data['Close'].pct_change(periods=20)  # Monthly momentum

# Handle zero volume before log transformation
data['Log Volume'] = np.log(data['Volume'].replace(0, 1e-10))  # Replace 0 with a small value
```


- added those to the scalar
```



# Feature scaling using StandardScaler
scaler = StandardScaler()
# Standardize the additional features and include in the model training
data[['Daily Returns', 'Volatility', 'Momentum', 'Log Volume']] = scaler.fit_transform(
    data[['Daily Returns', 'Volatility', 'Momentum', 'Log Volume']])
```
